File: mirror/tier2-dart/showcase/perf_caching.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Optimization
// Module: perf_caching.dart

class CachedFibonacci {
  final Map<int, int> _cache = {};
  int _hits = 0;
  int _misses = 0;
  
  int compute(int n) {
    if (_cache.containsKey(n)) {
      _hits++;
      return _cache[n]!;
    }
    
    _misses++;
    final result = n < 2 ? n : compute(n - 1) + compute(n - 2);
    _cache[n] = result;
    return result;
  }
  
  Map<String, dynamic> getStats() => {
    'hits': _hits,
    'misses': _misses,
    'cacheSize': _cache.length,
  };
}

void main() {
  final fib = CachedFibonacci();
  final fib10 = fib.compute(10);
  final fib20 = fib.compute(20);
  final stats = fib.getStats();
  
  final result = <String, dynamic>{
    'fib10': fib10,
    'fib20': fib20,
    'cacheHits': stats['hits'],
    'cacheMisses': stats['misses'],
    'cacheEnabled': true,
  };
  print(result);
}

File: mirror/tier2-python/showcase/perf_caching.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Optimization
Module: perf_caching.py
Purpose: Demonstrate caching and memoization
"""


class CachedFibonacci:
    """Fibonacci with caching"""
    def __init__(self):
        self.cache = {}
        self.hits = 0
        self.misses = 0
    
    def compute(self, n: int) -> int:
        if n in self.cache:
            self.hits += 1
            return self.cache[n]
        
        self.misses += 1
        if n < 2:
            result = n
        else:
            result = self.compute(n - 1) + self.compute(n - 2)
        
        self.cache[n] = result
        return result
    
    def get_stats(self):
        return {
            "hits": self.hits,
            "misses": self.misses,
            "cache_size": len(self.cache)
        }


fib = CachedFibonacci()
fib_10 = fib.compute(10)
fib_20 = fib.compute(20)
stats = fib.get_stats()

result = {
    "fib_10": fib_10,
    "fib_20": fib_20,
    "cache_hits": stats["hits"],
    "cache_misses": stats["misses"],
    "cache_enabled": True,
}

if __name__ == "__main__":
    print(result)
